tests/ship_test.py

```diff
@@ -0,0 +1,235 @@
+    specificExtraNodeosArgs[shipNodeNum]="--plugin eosio::state_history_plugin --disable-replay-opts --sync-fetch-span 200 --plugin eosio::net_api_plugin "
+
+    if cluster.launch(pnodes=totalProducerNodes,
+                      totalNodes=totalNodes, totalProducers=totalProducers,
+                      useBiosBootFile=False, specificExtraNodeosArgs=specificExtraNodeosArgs) is False:
+        Utils.cmdError("launcher")
+        Utils.errorExit("Failed to stand up eos cluster.")
+
+    # ***   identify each node (producers and non-producing node)   ***
+
+    shipNode = cluster.getNode(shipNodeNum)
+    prodNode = cluster.getNode(0)
+
+    #verify nodes are in sync and advancing
+    cluster.waitOnClusterSync(blockAdvancing=5)
+    Print("Cluster in Sync")
+
+    javascriptClient = "tests/ship_client.js"
+    cmd = "node %s --num-requests %d" % (javascriptClient, args.num_requests)
+    if Utils.Debug: Utils.Print("cmd: %s" % (cmd))
+    clients = []
+    files = []
+    shipTempDir = "%s/ship/" % (Utils.DataDir)
+    os.makedirs(shipTempDir, exist_ok = True)
+    shipClientFilePrefix = "%s/client" % (shipTempDir)
+
+    starts = []
+    for i in range(0, args.num_clients):
+        start = time.perf_counter()
+        outFile = open("%s%d.out" % (shipClientFilePrefix, i), "w")
+        errFile = open("%s%d.err" % (shipClientFilePrefix, i), "w")
+        Print("Start client %d" % (i))
+        popen=Utils.delayedCheckOutput(cmd, stdout=outFile, stderr=errFile)
+        starts.append(time.perf_counter())
+        clients.append((popen, cmd))
+        files.append((outFile, errFile))
+        Print("Client %d started, Ship node head is: %s" % (i, shipNode.getBlockNum()))
+
+    Print("Stopping all %d clients" % (args.num_clients))
+
+    index = 0
+    for popen, _ in clients:
+        popen.wait()
+        Print("Stopped client %d.  Ran for %s seconds." % (index, time.perf_counter() - starts[index]))
+        filePair = files[index]
+        filePair[0].close()
+        filePair[1].close()
+        index += 1
+
+    Print("Shutdown state_history_plugin nodeos")
+    shipNode.kill(signal.SIGTERM)
+
+    files = None
+
+    maxFirstBN = None
+    minLastBN = None
+    for index in range(0, len(clients)):
+        done = False
+        shipClientErrorFile = "%s%d.err" % (shipClientFilePrefix, i)
+        with open(shipClientErrorFile, "r") as errFile:
+            statuses = None
+            try:
+                statuses = json.load(errFile)
+            except json.decoder.JSONDecodeError as er:
+                Utils.errorExit("javascript client output was malformed in %s. Exception: %s" % (shipClientErrorFile, er))
+
+            for status in statuses:
+                statusDesc = status["status"]
+                if statusDesc == "done":
+                    done = True
+                    firstBlockNum = status["first_block_num"]
+                    lastBlockNum = status["last_block_num"]
+                    if maxFirstBN is None:
```

Comment:
No need for the conditional with the above initializations.  Initial values above will never be seen or used since an `error` status will result in `errorExit()` or `assert done` with `done` still equal to `False`.